### PR TITLE
Firmware command

### PR DESCRIPTION
**Pull request by mchack-work**
Replace SignPH() with GetFWDigest()

See https://github.com/tillitis/tkey-sign-cli/pull/21 for a way of testing this.

### CODE REVIEW

File: tkeysign.go [hunk 1]
```diff
@@ -297,11 +302,7 @@ func (s Signer) SignPh(data [64]byte) ([]byte, error) {
...
 
-	// Get the signature
-	signature, err := s.getSig()
-	if err != nil {
-		return nil, fmt.Errorf("getSig: %w", err)
-	}
-
-	return signature, nil
+	// Skip frame header, app header, and status; returning size of
+	// ed25519 signature
```
Comment (dehanj) [inline tkeysign.go]: "returning firmware digest", not ed25519 signature.

File: tkeysign.go [hunk 2]
```diff
@@ -297,11 +302,7 @@ func (s Signer) SignPh(data [64]byte) ([]byte, error) {
 		return nil, fmt.Errorf("SignData NOK")
```
Comment (dehanj) [inline tkeysign.go]: "GetFirmwareHash NOK"

File: tkeysign.go [hunk 3]
```diff
@@ -271,24 +271,29 @@ func (s Signer) getSig() ([]byte, error) {
 	return rx[3 : 3+64], nil
 }
 
-// SignPh signs a SHA512 pre-hashed message in data and returns an
-// ed25519ph signature.
-func (s Signer) SignPh(data [64]byte) ([]byte, error) {
+// GetFWDigest asks the signer to hash len bytes of the firmware.
```
Comment (dehanj) [inline tkeysign.go]: "asks the signer app ..."

**Review by dehanj — COMMENTED** [review]
Looks good, just minor comments.

File: tkeysign.go [hunk 4]
```diff
@@ -271,24 +271,29 @@ func (s Signer) getSig() ([]byte, error) {
...
-	tkeyclient.Dump("LoadSignPhData tx", tx)
+	tkeyclient.Dump("GetFirmwareHash tx", tx)
 	if err = s.tk.Write(tx); err != nil {
 		return nil, fmt.Errorf("Write: %w", err)
 	}
 
 	// Wait for reply
-	rx, _, err := s.tk.ReadFrame(rspSignPhData, id)
+	rx, _, err := s.tk.ReadFrame(rspGetFirmwareHash, id)
 	if err != nil {
```
Comment (dehanj) [inline tkeysign.go:297]: Potentially that one would want to have a dump of the rx content as well, when using --verbose flag.
`tkeyclient.Dump("GetFirmwareHash rx", rx)`

Comment (mchack-work) [reply tkeysign.go:297]: Sounds good. Added.

File: tkeysign.go [hunk 5]
```diff
@@ -271,37 +271,40 @@ func (s Signer) getSig() ([]byte, error) {
...
-	signature, err := s.getSig()
-	if err != nil {
-		return nil, fmt.Errorf("getSig: %w", err)
+	if rx[2] != tkeyclient.StatusOK {
+		return nil, fmt.Errorf("GetFirmwareHash NOK")
 	}
 
-	return signature, nil
+	// Skip frame header, app header, and status; returning
+	// firmware BLAKE2s digest
```
Comment (dehanj) [inline tkeysign.go]: The hash function used for hashing the firmware is sha512.

File: tkeysign.go [hunk 6]
```diff
@@ -271,37 +271,40 @@ func (s Signer) getSig() ([]byte, error) {
 	return rx[3 : 3+64], nil
 }
 
-// SignPh signs a SHA512 pre-hashed message in data and returns an
-// ed25519ph signature.
-func (s Signer) SignPh(data [64]byte) ([]byte, error) {
+// GetFWDigest asks the signer app to hash len bytes of the firmware.
+//
+// It returns the resulting BLAKE2s digest or an error.
```
Comment (dehanj) [inline tkeysign.go]: It returns the sha512 digest.

File: tkeysign.go (hunk 5 above)
Comment (dehanj) [reply tkeysign.go]: Fixed.

File: tkeysign.go (hunk 6 above)
Comment (dehanj) [reply tkeysign.go]: Fixed.

**Review by dehanj — APPROVED** [review]

**Pull request merged**